sdk/core/System.ClientModel/src/Pipeline/UserAgentPolicy.cs

```diff
@@ -0,0 +1,202 @@
+#endif
+
+        var platformInformation = EscapeProductInformation($"({RuntimeInformation.FrameworkDescription}; {osDescription})");
+
+        return applicationId != null
+            ? $"{applicationId} {assemblyName}/{version} {platformInformation}"
+            : $"{assemblyName}/{version} {platformInformation}";
+    }
+
+    /// <summary>
+    /// If the ProductInformation is not in the proper format, this escapes any ')' , '(' or '\' characters per https://www.rfc-editor.org/rfc/rfc7230#section-3.2.6
+    /// </summary>
+    /// <param name="productInfo">The ProductInfo portion of the user agent</param>
+    /// <returns>The escaped product information string.</returns>
+    private static string EscapeProductInformation(string productInfo)
+    {
+        // If the string is already valid, we don't need to escape anything
+        bool success = false;
+        try
+        {
+            success = ProductInfoHeaderValue.TryParse(productInfo, out var _);
+        }
+        catch (Exception)
+        {
+            // Invalid values can throw in Framework due to https://github.com/dotnet/runtime/issues/28558
+            // Treat this as a failure to parse.
+        }
+        if (success)
+        {
+            return productInfo;
+        }
+
+        var sb = new StringBuilder(productInfo.Length + 2);
+        sb.Append('(');
+        // exclude the first and last characters, which are the enclosing parentheses
+        for (int i = 1; i < productInfo.Length - 1; i++)
+        {
+            char c = productInfo[i];
+            if (c == ')' || c == '(')
+            {
+                sb.Append('\\');
+            }
+            // If we see a \, we don't need to escape it if it's followed by a '\', '(', or ')', because it is already escaped.
+            else if (c == '\\')
+            {
+                if (i + 1 < (productInfo.Length - 1))
+                {
+                    char next = productInfo[i + 1];
+                    if (next == '\\' || next == '(' || next == ')')
+                    {
+                        sb.Append(c);
+                        sb.Append(next);
+                        i++;
+                        continue;
+                    }
+                    else
+                    {
+                        sb.Append('\\');
+                    }
+                }
+                else
+                {
+                    sb.Append('\\');
+                }
+            }
+            sb.Append(c);
+        }
+        sb.Append(')');
+        return sb.ToString();
+    }
+
+    /// <summary>
+    /// Checks if a string contains any non-ASCII characters.
+    /// </summary>
+    /// <param name="value">The string to check.</param>
+    /// <returns>True if the string contains non-ASCII characters, false otherwise.</returns>
+    private static bool ContainsNonAscii(string value)
```

Comment:
do we have a target for .net 8? If yes, should we use: https://learn.microsoft.com/en-us/dotnet/api/system.text.ascii.isvalid?view=net-9.0#system-text-ascii-isvalid(system-readonlyspan((system-char)))